### PR TITLE
Add compiler driver based on pass manager

### PR DESCRIPTION
**Pull request by kffaxen**

### CODE REVIEW

File: src/Feldspar/Compiler/Compiler.hs [hunk 1]
```diff
@@ -154,3 +167,148 @@ sicsOptions2 = defaultOptions { frontendOpts = defaultFeldOpts { targets = [SICS
...
+programOpts :: SyntacticFeld a => a -> Options -> IO ()
+programOpts p opts = do args <- getArgs
+                        programOptsArgs p defaultProgOpts{backOpts = opts} args
+
+programOptsArgs :: SyntacticFeld a => a -> ProgOpts -> [String] -> IO ()
+programOptsArgs p opts args = do name <- getProgName
+                                 let opts0 = opts{passFileName = name ++ ".passes",
+                                                  outFileName = name,
+                                                  functionName = name}
+                                 let (opts1,_) = decodeOpts opts0 args
```
Comment (pjonsson) [inline src/Feldspar/Compiler/Compiler.hs]: Make decodeOpts return a single thing instead. And I think the bound variables can reside in the same let.

File: src/Feldspar/Compiler/Compiler.hs [hunk 2]
```diff
@@ -154,3 +167,148 @@ sicsOptions2 = defaultOptions { frontendOpts = defaultFeldOpts { targets = [SICS
...
+  ]
+
+chooseEnd :: (forall a . PassCtrl a -> a -> PassCtrl a) -> String -> ProgOpts -> ProgOpts
+chooseEnd f str opts
+  | [(p,_)] <- reads str = opts{frontendCtrl = f (frontendCtrl opts) p}
+  | [(p,_)] <- reads str = opts{backendCtrl = f (backendCtrl opts) p}
+  | otherwise = error $ "Compiler.chooseEnd: unrecognized pass " ++ str
+
+writeFileLB :: String -> String -> IO ()
+writeFileLB   "" str = putStr str
```
Comment (pjonsson) [inline src/Feldspar/Compiler/Compiler.hs]: Why not use - for stdout?

File: src/Feldspar/Compiler/Compiler.hs (hunk 1 above)
Comment (kffaxen) [reply src/Feldspar/Compiler/Compiler.hs]: Now the non-options are also used.

File: src/Feldspar/Compiler/Compiler.hs (hunk 2 above)
Comment (kffaxen) [reply src/Feldspar/Compiler/Compiler.hs]: Yes, that is better.

**Pull request merged**